src/IslandoraUtils.php

```diff
@@ -227,28 +231,64 @@ public function getReferencingMedia($fid) {
     // Get media fields that reference files.
     $fields = $this->getReferencingFields('media', 'file');
 
+    if (empty($fields)) {
+      // No referencing fields, nothing able to be returned.
+      return [];
+    }
+
     // Process field names, stripping off 'media.' and appending 'target_id'.
     $conditions = array_map(
-      function ($field) {
+      static function (string $field) {
         if (str_starts_with($field, 'media.')) {
           $field = substr($field, 6);
         }
-        return $field . '.target_id';
+        return $field . '_target_id';
       },
       $fields
     );
 
-    // Query for media that reference this file.
-    $query = $this->entityTypeManager->getStorage('media')->getQuery();
-    $query->accessCheck(TRUE);
-    $group = $query->orConditionGroup();
-    foreach ($conditions as $condition) {
-      $group->condition($condition, $fid);
+    assert(count($conditions) > 0);
+
+    /** @var \Drupal\Core\Database\Query\SelectInterface[] $queries */
+    $queries = [];
+    foreach ($conditions as $key => $field) {
+      /* Encode the query:
+       *
+       * @code
+       * SELECT media__field_media_file.revision_id
+       * FROM media__field_media_file media__field_media_file
+       * WHERE media__field_media_file.field_media_file_target_id = $fid
+       * @endcode
+       */
+      $table = str_replace('.', '__', $key);
+      $subquery = $this->database->select($table, $table);
+      $subquery->addField($table, 'revision_id', 'vid');
+      $subquery->condition("{$table}.{$field}", $fid, '=');
+
+      // SELECT base_table.vid AS vid, base_table.mid AS mid
+      // FROM media base_table WHERE base_table.vid IN.
+      $queries[$key] = $this->database->select('media', 'm')
+        ->fields('m', ['mid'])
+        ->condition('m.vid', $subquery, 'IN');
     }
-    $query->condition($group);
 
-    return $this->entityTypeManager->getStorage('media')
-      ->loadMultiple($query->execute());
+    assert(count($queries) > 0);
+
+    $unionQuery = array_shift($queries);
+    foreach ($queries as $queryPart) {
+      $unionQuery->union($queryPart);
```

Comment:
```suggestion
      $unionQuery->union($queryPart, 'ALL');
```

Might work better for MySQL/MariaDB, given https://mariadb.com/docs/server/reference/sql-statements/data-manipulation/selecting-data/joins-subqueries/union : 

> The server can in most cases execute UNION ALL without creating a temporary table

Drupal method, just for reference: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Database%21Query%21Select.php/function/Select%3A%3Aunion/10

Reply:
I tried adding this suggestion after applying the patch and it resulted in the same ~40s execution time